pkg/operation/botanist/controlplane/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,561 @@
+	}
+	rb, err := r.emptyRoleBindingInWatchedNamespace()
+	if err == nil {
+		objectsToDelete = append(objectsToDelete, rb)
+	}
+
+	return kutil.DeleteObjects(
+		ctx,
+		r.client,
+		objectsToDelete...,
+	)
+}
+
+func (r *resourceManager) ensureRBAC(ctx context.Context) error {
+	targetDiffersFromSourceCluster := r.values.Kubeconfig != nil
+	if targetDiffersFromSourceCluster {
+		if r.values.WatchedNamespace == nil {
+			if err := r.ensureClusterRole(ctx, allowManagedResources); err != nil {
+				return err
+			}
+			if err := r.ensureClusterRoleBinding(ctx); err != nil {
+				return err
+			}
+		} else {
+			if err := r.ensureRoleInWatchedNamespace(ctx, allowManagedResources); err != nil {
+				return err
+			}
+			if err := r.ensureRoleBinding(ctx); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := r.ensureClusterRole(ctx, allowAll); err != nil {
+			return err
+		}
+		if err := r.ensureClusterRoleBinding(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *resourceManager) ensureClusterRole(ctx context.Context, policies []rbacv1.PolicyRule) error {
+	clusterRole := r.emptyClusterRole()
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, clusterRole, func() error {
+		clusterRole.Labels = r.getLabels()
+		clusterRole.Rules = policies
+		return nil
+	})
+	return err
+}
+
+func (r *resourceManager) emptyClusterRole() *rbacv1.ClusterRole {
+	return &rbacv1.ClusterRole{ObjectMeta: metav1.ObjectMeta{Name: clusterRoleName}}
+}
+
+func (r *resourceManager) ensureClusterRoleBinding(ctx context.Context) error {
+	clusterRoleBinding := r.emptyClusterRoleBinding()
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, clusterRoleBinding, func() error {
+		clusterRoleBinding.Labels = r.getLabels()
+		clusterRoleBinding.RoleRef = rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     clusterRoleName,
+		}
+		clusterRoleBinding.Subjects = []rbacv1.Subject{{
+			Kind:      rbacv1.ServiceAccountKind,
+			Name:      serviceAccountName,
+			Namespace: r.namespace,
+		}}
+		return nil
+	})
+	return err
+}
+
+func (r *resourceManager) emptyClusterRoleBinding() *rbacv1.ClusterRoleBinding {
+	return &rbacv1.ClusterRoleBinding{ObjectMeta: metav1.ObjectMeta{Name: clusterRoleName}}
+}
+
+func (r *resourceManager) ensureRoleInWatchedNamespace(ctx context.Context, policies []rbacv1.PolicyRule) error {
+	role, err := r.emptyRoleInWatchedNamespace()
+	if err != nil {
+		return err
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, role, func() error {
+		role.Labels = r.getLabels()
+		role.Rules = policies
+		return nil
+	})
+	return err
+}
+
+func (r *resourceManager) emptyRoleInWatchedNamespace() (*rbacv1.Role, error) {
+	if r.values.WatchedNamespace == nil {
+		return nil, errors.New("creating Role in watched namespace failed - no namespace defined")
+	}
+	return &rbacv1.Role{ObjectMeta: metav1.ObjectMeta{Name: roleName, Namespace: *r.values.WatchedNamespace}}, nil
+}
```

Comment:
Hm, is it really necessary to return an error here? I don't see that this would a benefit, or? Can't we simply return `nil` and ease our life here?

Reply:
Mmh, isn't it closer to what you would expect in go to return an error instead of nil for the object? Even if I switch to nil I would still need to perform the checks when I call this function but I would have to check whether the returned role is `nil`. I feel we would not gain much from that.

Reply:
I see your point that this error feels kind of unnecessary but I am not sure how to improve it right now.

Reply:
Well, no strong opinion and no need to block anything here. Resolving the comment for now.